RecoMuon/SimMuonAlgos/plugins/MuonSimClassifier.cc

```diff
@@ -0,0 +1,574 @@
+// -*- C++ -*-
+//
+// Package:    MuonSimClassifier
+// Class:      MuonSimClassifier
+//
+/**\class MuonSimClassifier MuonSimClassifier.cc MuonAnalysis/MuonAssociators/src/MuonSimClassifier.cc
+
+
+ CLASSIFICATION: For each RECO Muon, match to SIM particle, and then:
+  - If the SIM is not a Muon, label as Punchthrough (1) except if it is an electron or positron (11)
+  - If the SIM is a Muon, then look at it's provenance.
+     A) the SIM muon is also a GEN muon, whose parent is NOT A HADRON AND NOT A TAU
+        -> classify as "primary" (4).
+     B) the SIM muon is also a GEN muon, whose parent is HEAVY FLAVOURED HADRON OR A TAU
+        -> classify as "heavy flavour" (3)
+     C) classify as "light flavour/decay" (2)
+
+  In any case, if the TP is not preferentially matched back to the same RECO muon,
+  label as Ghost (flip the classification)
+
+
+ FLAVOUR:
+  - for non-muons: 0
+  - for primary muons: 13
+  - for non primary muons: flavour of the mother: std::abs(pdgId) of heaviest quark, or 15 for tau
+
+*/
+//
+// Original Author:  G.Petrucciani and G.Abbiendi
+//         Created:  Sun Nov 16 16:14:09 CET 2008
+//         revised:  3/Aug/2017
+//
+
+
+// system include files
+#include <memory>
+#include <set>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "DataFormats/Common/interface/View.h"
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "DataFormats/Common/interface/Association.h"
+
+#include "DataFormats/MuonReco/interface/Muon.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticleFwd.h"
+
+#include "SimDataFormats/Associations/interface/MuonToTrackingParticleAssociator.h"
+#include "SimTracker/Records/interface/TrackAssociatorRecord.h"
+#include <SimDataFormats/TrackingAnalysis/interface/TrackingVertex.h>
+
+#include "CommonTools/Utils/interface/StringCutObjectSelector.h"
+
+#include <boost/foreach.hpp>
+#define foreach BOOST_FOREACH
```

Comment:
this is not used, apparently.

Reply:
Since I removed some code that can be true. Will check it.

Reply:
fixed